### ISSUE TITLE
SMB remote returns different file list than e.g. `smbclient`

### ISSUE THREAD

**Issue by StanleyP** [rclone/rclone#6672]
#### What is the problem you are having with rclone?

I am accessing SMB file share on Windows server and rclone gives me different file list than when I list the files using for instance linux `smbclient`.

I found that the issues is in `go-smb2` library that Rclone is using for SMB functionality and has relation to following issue:

https://github.com/hirochachacha/go-smb2/issues/40

The author of `go-smb2` library acknoledges the problem, but hesitates to fix it, because it implies breaking change in the library public API. (The library does not have access to original hostname).

I created patch based on suggestions and attempts from the issue above that fixes the problem. So I think that either somebody from Rclone community might find it useful or evetually it could be integrated into the Rclone itself somehow.

Please excuse me because the patch is quite "raw", I'm not a go coder, I just wanted to make it work. You can surely find the better way how to integrate this change :)

##### File `go-smb2.patch`

```
diff --git a/client.go b/client.go
index e0f7b36..f0ad2e3 100644
--- a/client.go
+++ b/client.go
@@ -30,8 +30,8 @@ type Dialer struct {
 // It returns a session. It doesn't support NetBIOS transport.
 // This implementation doesn't support multi-session on the same TCP connection.
 // If you want to use another session, you need to prepare another TCP connection at first.
-func (d *Dialer) Dial(tcpConn net.Conn) (*Session, error) {
-	return d.DialContext(context.Background(), tcpConn)
+func (d *Dialer) Dial(tcpConn net.Conn, host string) (*Session, error) {
+	return d.DialContext(context.Background(), tcpConn, host)
 }
 
 // DialContext performs negotiation and authentication using the provided context.
@@ -39,7 +39,7 @@ func (d *Dialer) Dial(tcpConn net.Conn) (*Session, error) {
 // If you want to use the same context, call Session.WithContext manually.
 // This implementation doesn't support multi-session on the same TCP connection.
 // If you want to use another session, you need to prepare another TCP connection at first.
-func (d *Dialer) DialContext(ctx context.Context, tcpConn net.Conn) (*Session, error) {
+func (d *Dialer) DialContext(ctx context.Context, tcpConn net.Conn, host string) (*Session, error) {
 	if ctx == nil {
 		panic("nil context")
 	}
@@ -69,13 +69,14 @@ func (d *Dialer) DialContext(ctx context.Context, tcpConn net.Conn) (*Session, e
 		return nil, err
 	}
 
-	return &Session{s: s, ctx: context.Background(), addr: tcpConn.RemoteAddr().String()}, nil
+	return &Session{s: s, ctx: context.Background(), host: host, addr: tcpConn.RemoteAddr().String()}, nil
 }
 
 // Session represents a SMB session.
 type Session struct {
 	s    *session
 	ctx  context.Context
+	host string
 	addr string
 }
 
@@ -83,7 +84,7 @@ func (c *Session) WithContext(ctx context.Context) *Session {
 	if ctx == nil {
 		panic("nil context")
 	}
-	return &Session{s: c.s, ctx: ctx, addr: c.addr}
+	return &Session{s: c.s, ctx: ctx, host: c.host, addr: c.addr}
 }
 
 // Logoff invalidates the current SMB session.
@@ -99,7 +100,11 @@ func (c *Session) Mount(sharename string) (*Share, error) {
 	sharename = normPath(sharename)
 
 	if !strings.ContainsRune(sharename, '\\') {
-		sharename = fmt.Sprintf(`\\%s\%s`, c.addr, sharename)
+		if c.host != "" {
+			sharename = fmt.Sprintf(`\\%s\%s`, c.host, sharename)
+		} else {
+			sharename = fmt.Sprintf(`\\%s\%s`, c.addr, sharename)
+		}
 	}
 
 	if err := validateMountPath(sharename); err != nil {
@@ -116,6 +121,9 @@ func (c *Session) Mount(sharename string) (*Share, error) {
 
 func (c *Session) ListSharenames() ([]string, error) {
 	servername := c.addr
+	if c.host != "" {
+		servername = c.host
+	}
 
 	fs, err := c.Mount(fmt.Sprintf(`\\%s\IPC$`, servername))
 	if err != nil {
diff --git a/session.go b/session.go
index 0292f9e..5feade5 100644
--- a/session.go
+++ b/session.go
@@ -265,6 +265,8 @@ type session struct {
 	decrypter cipher.AEAD
 
 	// applicationKey []byte
+
+	serverName string
 }
 
 func (s *session) logoff(ctx context.Context) error {
```

##### File `rclone.patch`

```
diff --git a/backend/smb/connpool.go b/backend/smb/connpool.go
index d4bed787c..f4b45a984 100644
--- a/backend/smb/connpool.go
+++ b/backend/smb/connpool.go
@@ -40,7 +40,7 @@ func (f *Fs) dial(ctx context.Context, network, addr string) (*conn, error) {
 		},
 	}
 
-	session, err := d.DialContext(ctx, tconn)
+	session, err := d.DialContext(ctx, tconn, addr)
 	if err != nil {
 		return nil, err
 	}
```

##### File `Dockerfile`

```
FROM golang

COPY *.patch /usr/src/
RUN set -o errexit \
  && mkdir -p /go/src/github.com/rclone \
  && git clone -b v1.61.1 https://github.com/rclone/rclone.git /go/src/github.com/rclone/rclone
WORKDIR /go/src/github.com/rclone/rclone/

RUN set -o errexit \
  && git apply < /usr/src/rclone.patch \
  && go get \
  && ( set -o errexit; \
       cd /go/pkg/mod/github.com/hirochachacha/go-smb2@v1.1.0; \
       git apply < /usr/src/go-smb2.patch \
     ) \
  && CGO_ENABLED=0 \
  make
RUN ./rclone version
```

#### What is your rclone version (output from `rclone version`)

`1.61.1`

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

`Ubuntu 20.04, 64bit`

#### Which cloud storage system are you using? (e.g. Google Drive)

`smb`

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Good work finding a fix!

This really needs to go upstream.

I posted on the thread with an idea.

**Comment by arashpayan** [comment]
Hi @ncw. I'm the maintainer of the go-smb2 fork at github.com/cloudsoda/go-smb2. I've fixed this issue in a local branch, and will be bringing it into `main` later today or tomorrow.

In the interest of making the 'right' way of using the library the only of using the library, I want to replace the existing `Dial(net.Conn)` and `DialContext(context.Context, net.Conn)` methods with a single `Dial(ctx context.Context, addr string)`. The new method will handle the tcp connection creation. Would removing the ability to provide a `net.Conn` be a problem for rclone?

**Comment by ncw** [comment]
@arashpayan 
> I'm the maintainer of the go-smb2 fork at github.com/cloudsoda/go-smb2. I've fixed this issue in a local branch, and will be bringing it into `main` later today or tomorrow.

Great!

> In the interest of making the 'right' way of using the library the only of using the library, I want to replace the existing `Dial(net.Conn)` and `DialContext(context.Context, net.Conn)` methods with a single `Dial(ctx context.Context, addr string)`. The new method will handle the tcp connection creation. Would removing the ability to provide a `net.Conn` be a problem for rclone?

Yes... Rclone has its own Dialer (created with `fshttp.NewDialer`) which does all sorts of great stuff (like set QoS, do rate limiting etc). It wouldn't be the end of the world to lose this but it would be nice to keep it.

You could fix this by allowing a config parameter for the user to supply their own `DialContext` in `smb2.Dialer`

    DialContext(ctx context.Context, network, address string) (net.Conn, error)

Here is the code in rclone so you can take a look

https://github.com/rclone/rclone/blob/668711e43213bc9464c0b25c88621cd987898f08/backend/smb/connpool.go#L16-L19

PS as far as rclone goes, please feel free to break the API! As long as it fails to compile I don't care. If you break the API and it still compiles then I probably won't notice!

**Comment by arashpayan** [comment]
Interesting. Those look like worthwhile settings in `fshttp`. I'll make sure to provide a way for rclone to keep leveraging that.

**Comment by halms** [comment]
@arashpayan, I just tried your updated `go-smb2` with rclone and am really happy!
Just required minimal changes in rclone and finally resolved my issue accessing a AWS FSx Windows file share.

Do you want to prepare a PR for rclone to use the new version or should I do that?

**Comment by arashpayan** [comment]
:wave: @halms. @ncw already switched rclone over to using github.com/cloudsoda/go-smb2 a few days ago. I think you just need to wait for a release now. :slightly_smiling_face:

**Comment by halms** [comment]
Yes, but it's currently on a [previous version](https://github.com/rclone/rclone/blob/96f8b7c827261200dad7d7d786f6778cb87e0992/go.mod#L23), and for the latest, `connpool.go` needs a small change.
https://github.com/rclone/rclone/blob/96f8b7c827261200dad7d7d786f6778cb87e0992/backend/smb/connpool.go#L43
```diff
-	session, err := d.DialContext(ctx, tconn)
+	session, err := d.DialConn(ctx, tconn, addr)
```

**Comment by halms** [comment]
[PR](https://github.com/rclone/rclone/pull/7465) opened.

I don't know what it exactly is (I don't have any insight how this FSx was setup, I'm just a consumer in that regard).
From what I see, it is a drive joined to an AD with an internal domain name `amznfsx···.xxx.local` in a private network.

Overall it just is as described here / in [your issue](https://github.com/CloudSoda/go-smb2/issues/18):
When the IP address is used in the server part of the share the the share is not available, when the correct hostname is used, then it works.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.66 - I'll probably release it in v1.64.1 also.

Thank you @halms  and @arashpayan
